### PR TITLE
State/block pruning

### PR DESCRIPTION
**Pull request by yglukhov**

### CODE REVIEW

**Comment by arnetheduck** [comment]
so this is the perfect branch to start adding some tests - there are some basic ones in `test_block_pool.nim` already

File: beacon_chain/block_pool.nim [hunk 1]
```diff
@@ -421,6 +431,45 @@ proc loadTailState*(pool: BlockPool): StateData =
...
+  # (except the most recent)
+  if (let blk = pool.db.getBlock(b.root); blk.isSome):
+    pool.db.delState(blk.get.stateRoot)
+
+proc setTailBlock(pool: BlockPool, newTail: BlockRef) =
+  ## Advance tail block, pruning all the states and blocks with older slots
+  let oldTail = pool.tail
+  let fromSlot = oldTail.slot.uint64
+  let toSlot = newTail.slot.uint64 - 1
+  assert(toSlot > fromSlot)
```
Comment (arnetheduck) [inline beacon_chain/block_pool.nim]: maybe better just do nothing in this case, there's no need to be harsh

File: beacon_chain/block_pool.nim [hunk 2]
```diff
@@ -421,6 +431,45 @@ proc loadTailState*(pool: BlockPool): StateData =
...
+  for s in fromSlot .. toSlot:
+    for b in pool.blocksBySlot.getOrDefault(s, @[]):
+      if (let blko = pool.db.getBlock(b.root); blko.isSome):
+        let stateRoot = blko.get().stateRoot
+        pool.db.delState(stateRoot)
+        pool.db.delBlock(b.root)
+
+        b.children = @[]
```
Comment (arnetheduck) [inline beacon_chain/block_pool.nim]: the cleanup here is similar to the cleanup when dropping a branch block.. join the two?

File: beacon_chain/block_pool.nim (hunk 1 above)
Comment (yglukhov) [reply beacon_chain/block_pool.nim]: This function is private and there's no reason why this function would be called with `newTail` older than `oldTail`. Thus failure of this assertion would indicate a bug in `newTail` calculation.

**Comment by tersec** [comment]
Is this still in progress?

**Comment by arnetheduck** [comment]
tests pending

**Comment by tersec** [comment]
?

**Comment by arnetheduck** [comment]
waiting for tests, so that we have the scaffolding for then and it's easy to add more

**Comment by tersec** [comment]
?

**Pull request merged**